qiskit_machine_learning/circuit/library/qnn_circuit.py

```diff
@@ -0,0 +1,217 @@
+        self.num_qubits, self._feature_map, self._ansatz = derive_num_qubits_feature_map_ansatz(
+            num_qubits, feature_map, ansatz
+        )
+
+    def _build(self):
+        super()._build()
+        self.compose(self.feature_map, inplace=True)
+        self.compose(self.ansatz, inplace=True)
+
+    def _check_configuration(self, raise_on_failure=True):
+        try:
+            self.num_qubits, self.feature_map, self.ansatz = derive_num_qubits_feature_map_ansatz(
+                self.num_qubits, self.feature_map, self.ansatz
+            )
+        except QiskitMachineLearningError as qml_ex:
+            if raise_on_failure:
+                raise qml_ex
+
+    @property
+    def num_qubits(self) -> int:
+        """Returns the number of qubits in this circuit.
+
+        Returns:
+            The number of qubits.
+        """
+        return super().num_qubits
+
+    @num_qubits.setter
+    def num_qubits(self, num_qubits: int) -> None:
+        """Set the ``QNNCircuit`` number of qubits. If the ``QNNCircuit`` number of qubits differ
+        from the feature map or the ansatz number of qubits, these properties are adjusted
+        respectively.
+
+        Args:
+            num_qubits:  The number of qubits.
+        """
+        if self.num_qubits != num_qubits:
+            # invalidate the circuit
+            self._invalidate()
+            self.qregs: List[QuantumRegister] = []
+            if num_qubits is not None and num_qubits > 0:
+                self.qregs = [QuantumRegister(num_qubits, name="q")]
+                (
+                    self.num_qubits,
+                    self._feature_map,
+                    self._ansatz,
+                ) = derive_num_qubits_feature_map_ansatz(
+                    num_qubits, self._feature_map, self._ansatz
+                )
+
+    @property
+    def feature_map(self) -> QuantumCircuit:
+        """Returns feature_map.
+
+        Returns:
+            The feature map.
+        """
+        return self._feature_map
+
+    @feature_map.setter
+    def feature_map(self, feature_map: QuantumCircuit) -> None:
+        """Set the feature map. If the feature map number of qubits differ from the ``QNNCircuit``
+        or the ansatz number of qubits, these properties are adjusted respectively.
+
+        Args:
+            feature_map: The feature map.
+        """
+        if self.feature_map != feature_map:
+            # invalidate the circuit
+            self._invalidate()
+            self.num_qubits = feature_map.num_qubits
+            self.num_qubits, self._feature_map, self._ansatz = derive_num_qubits_feature_map_ansatz(
+                self.num_qubits, feature_map, self.ansatz
+            )
+
+    @property
+    def ansatz(self) -> QuantumCircuit:
+        """Returns ansatz.
+
+        Returns:
+            The ansatz.
+        """
+        return self._ansatz
+
+    @ansatz.setter
+    def ansatz(self, ansatz: QuantumCircuit) -> None:
```

Comment:
> these properties are adjusted respectively

I think this should be explained as to what this adjustment actually is.

As far as I can tell the QNNCircuit is updated to now have an overall num_qubits the same as the ansatz being passed to the setter (assuming QNNCircuits FeatureMap can be adjusted). 
```
  qc = QNNCircuit(feature_map=ZZFeatureMap(2), ansatz=RealAmplitudes(2))
  qc.ansatz=EfficientSU2(3, reps=1)
  print(qc)
```
Circuit is 3 qubits wide now having started out as 2.

In other words the circuit is adjusted to the ansatz being passed in, rather than the other way around.

Since it's the former, the ansatz passed has to have a positive number for num_qubits. The following, just passing in the ansatz without that being set fails (as it does not adjust the ansatz to the present circuit size)

```
  qc = QNNCircuit(feature_map=ZZFeatureMap(2), ansatz=RealAmplitudes(2))
  qc.ansatz=EfficientSU2(reps=1)
```
if you want that you can achieve that net result by setting it explicitly to match yourself
```
  qc = QNNCircuit(feature_map=ZZFeatureMap(2), ansatz=RealAmplitudes(2))
  qc.ansatz=EfficientSU2(qc.num_qubits, reps=1)
```

The FeatureMap setter has the same behavior as far as I can tell and its docstring should be improved in a similar manner.